client/rest-high-level/src/main/java/org/elasticsearch/client/ml/PostDataRequest.java

```diff
@@ -0,0 +1,224 @@
+     */
+    public static PostDataRequest postJsonDataRequest(String jobId) {
+        return new PostDataRequest(jobId, XContentType.JSON);
+    }
+
+    /**
+     * PostDataRequest for sending data in SMILE format
+     * @param jobId non-null jobId of the job to post data to
+     */
+    public static PostDataRequest postSmileDataRequest(String jobId) {
+        return new PostDataRequest(jobId, XContentType.SMILE);
+    }
+
+    /**
+     * Create a new PostDataRequest object
+     *
+     * @param jobId non-null jobId of the job to post data to
+     * @param xContentType content type of the data to post. Only {@link XContentType#JSON} {@link XContentType#SMILE} are supported
+     */
+    public PostDataRequest(String jobId, XContentType xContentType) {
+        this.jobId = Objects.requireNonNull(jobId, "job_id must not be null");
+        this.xContentType = Objects.requireNonNull(xContentType, "content_type must not be null");
+        this.bytesReferences = new ArrayList<>();
+        this.objectMaps = new ArrayList<>();
+    }
+
+    public String getJobId() {
+        return jobId;
+    }
+
+    public String getResetStart() {
+        return resetStart;
+    }
+
+    /**
+     * Specifies the start of the bucket resetting range
+     *
+     * @param resetStart String representation of a timestamp; may be an epoch seconds, epoch millis or an ISO string
+     */
+    public void setResetStart(String resetStart) {
+        this.resetStart = resetStart;
+    }
+
+    public String getResetEnd() {
+        return resetEnd;
+    }
+
+    /**
+     * Specifies the end of the bucket resetting range
+     *
+     * @param resetEnd String representation of a timestamp; may be an epoch seconds, epoch millis or an ISO string
+     */
+    public void setResetEnd(String resetEnd) {
+        this.resetEnd = resetEnd;
+    }
+
+    /**
+     * Gets the transformed content to post.
+     *
+     * This combines both documents added through {@link PostDataRequest#addDoc(BytesReference)} and {@link PostDataRequest#addDoc(Map)}
+     * into a single BytesReference object according to the set XContentType for bulk submission
+     *
+     * If content was set via {@link PostDataRequest#setContent(BytesReference)}, then simply that content is returned.
+     *
+     * @throws IOException on parsing/serialization errors
+     */
+    public BytesReference getContent() throws IOException {
+        if (content != null) {
+            return content;
+        }
+        try (XContentBuilder builder = XContentBuilder.builder(xContentType.xContent())) {
+            for (BytesReference bytesReference : bytesReferences) {
+                try (StreamInput streamInput = bytesReference.streamInput()) {
+                    builder.rawValue(streamInput, xContentType);
+                }
+            }
+            for (Map<String, Object> objectMap : objectMaps) {
+                builder.map(objectMap);
+            }
```

Comment:
There's a problem here because if data has been added as a mixture of bytes references and maps then all the maps will be sent after all the bytes references, and that could mean the data is not sent in ascending time order.

Two possible solutions I can think of are:

1. Document that all data must be supplied in the same format - either bytes references or maps - and enforce this in the `addDoc()` methods.
2. Convert the maps to bytes references in `addDoc()` so that there are only bytes references being stored.

This also shows that we should say somewhere in the Javadocs that docs will be processed by the job in the order they're added to the request and that therefore they should be added to the request in ascending time order.